edb/lib/ext/pgvector.edgeql

```diff
@@ -0,0 +1,189 @@
+        set sql_type_scheme := "vector({__arg_0__})";
+        set num_params := 1;
+    };
+
+    create cast from ext::pgvector::vector to std::str {
+        set volatility := 'Immutable';
+        using sql cast;
+    };
+
+    create cast from std::str to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+    };
+
+    create cast from ext::pgvector::vector to std::json {
+        set volatility := 'Immutable';
+        using sql 'SELECT val::text::jsonb';
+    };
+
+    create cast from std::json to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT (
+            nullif(val, 'null'::jsonb)::text::vector
+        )
+        $$;
+    };
+
+    # All casts from numerical arrays should allow assignment casts.
+    create cast from array<std::float32> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::float64> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::int16> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::float4[]::vector
+        $$;
+        allow assignment;
+    };
+
+    create cast from array<std::int32> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::int64> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::float4[]::vector
+        $$;
+        allow assignment;
+    };
+
+    create cast from array<std::decimal> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::bigint> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::numeric[]::vector
+        $$;
+        allow assignment;
+    };
+
+    create cast from ext::pgvector::vector to array<std::float32> {
+        set volatility := 'Immutable';
+        using sql cast;
+    };
+
+    create function ext::pgvector::euclidean_distance(
+        a: ext::pgvector::vector,
+        b: ext::pgvector::vector,
+    ) -> std::float64 {
+        set volatility := 'Immutable';
+        # Needed to pick up the indexes when used in ORDER BY.
+        set prefer_subquery_args := true;
+        using sql 'SELECT a <-> b';
+    };
+
+    create function ext::pgvector::inner_product(
+        a: ext::pgvector::vector,
+        b: ext::pgvector::vector,
+    ) -> std::float64 {
+        set volatility := 'Immutable';
+        # Needed to pick up the indexes when used in ORDER BY.
+        set prefer_subquery_args := true;
+        using sql 'SELECT -(a <#> b)';
```

Comment:
Hmm, so from pgvector README:

> Note: `<#>` returns the negative inner product since Postgres only supports ASC order index scans on operators

Why are we reverting this?  Have you checked that `ORDER BY inner_product(a, b) DESC` still uses the index?

Reply:
No, I was sloppy and only checked euclidean distance.

I generalized the test and sure enough it didn't work, so I turned inner_product into neg_inner_product